### PR TITLE
Allow contiguous leader across terms

### PR DESCRIPTION
**Pull request by carrino**
---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1262)

### CODE REVIEW

File: atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/TimeLockServer.java [hunk 1]
```diff
@@ -61,7 +63,30 @@ public void run(TimeLockServerConfiguration configuration, Environment environme
         LocalMember localMember = Futures.getUnchecked(timeLockGroup.join());
 
         DistributedValue<String> leaderId = DistributedValues.getLeaderId(localNode);
-        timeLockGroup.election().onElection(term -> Futures.getUnchecked(leaderId.set(term.leader().id())));
+        timeLockGroup.election().onElection(term -> {
+            LeaderAndTerm newLeader = ImmutableLeaderAndTerm.of(term.term(), term.leader().id());
+            while (true) {
+                String curValue = Futures.getUnchecked(leaderId.get());
+                LeaderAndTerm curLeader = LeaderAndTerm.fromStoredString(curValue);
+                if (curValue == null) {
```
Comment (schlosna) [inline atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/TimeLockServer.java]: should this be `if (curLeader == null) {`?

File: atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/TimeLockServer.java [hunk 2]
```diff
@@ -61,7 +63,30 @@ public void run(TimeLockServerConfiguration configuration, Environment environme
...
+                    // We have never had a leader so we should try to set it.
```
Comment (schlosna) [inline atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/TimeLockServer.java]: we probably want debug logging for every one of these comments for diagnostics at runtime.

**Comment by SerialVelocity** [comment]
This is such a small case that I'm not sure we even want to deal with this (if it can even happen at all). The added complexity isn't worth it.

**Comment by ThomasBoam** [comment]
@SerialVelocity can you explain why you think this is unlikely to happen?  If it does happen, what are the consequences?  Even if it is rare we should probably guard against the case.  Alternatively, can we create a test that proves this case cannot happen?

**Comment by SerialVelocity** [comment]
Sure.

The only case that @jeremyk-91 and I think this can happen is if a leader gets partitioned off from the other nodes then just as they think they need a new leader, it reconnects and becomes the "new" leader. (It may not even be able to happen at this point because the election has already started so maybe new nodes cannot be added). It's an extremely rare case.

If it does happen, we end up clearing the locks when they didn't need to be cleared. This may abort some transactions but should be totally safe.

**Comment by schlosna** [comment]
Defer to @carrino on what prompted this PR, but I think you need to try on real data for frequency of occurance in real workloads and the actual costs of the retries and fallbacks

**Comment by SerialVelocity** [comment]
I did implement it this way to start with but it seemed very complex. It seems that, since many people confused as to whether it works or not (because it relies on atomix internal guarantees), it may be better to implement it like this anyway.

**Comment by carrino** [comment]
I think this change is pretty straightforward and takes the system from
"mostly correct" to "correct".

I'd also like to add that we shouldn't be in the habit of saying stuff like
"this is a correctness bug but it won't happen much".  I've dismissed other
public projects for much less.

**Comment by SerialVelocity** [comment]
I meant that this makes it easier to reason about as well as improving perf for a very rare case. Clearing the locks too many times isn't a correctness bug afaik

**Comment by carrino** [comment]
Agreed that clearing locks is a perfect bug but since that code is a client
side trigger I don't think it's sufficient for fencing at request time

**Comment by ThomasBoam** [comment]
Definitely agree that we need to make this correct above all else

If this was purely a performance concern then we should try to measure the
improvement before adding complexity but if we're at all worried about
correctness then we should address the problem now.

This is only really a couple of small changes so it doesn't seem like a lot
of complexity IMO.

-Tom

On Wed, 23 Nov 2016, 17:25 Ben Grabham, <notifications@github.com> wrote:

I meant that this makes it easier to reason about as well as improving perf
for a very rare case. Clearing the locks too many times isn't a correctness
bug afaik

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/palantir/atlasdb/pull/1262#issuecomment-262579396>, or mute
the thread

**Comment by SerialVelocity** [comment]
It's easier to reason about with this version so we should switch to this anyway.

**Comment by carrino** [comment]
Probably. I don't know the atomix api. Do we need to add classes/jars to
their server to serialize stuff?

My change is more of a sketch of what I was thinking would work

**Comment by carrino** [comment]
I assume merging this change means the term doesn't churn unless leadership is changed?

**Comment by SerialVelocity** [comment]
Correct

**Comment by tpetracca** [comment]
Was thinking about this. Can we not just do what we've always done for timestamp service and instead of keeping info about leader and term just store a local version of what we believe the current value of the distributed long and any time it's not that value anymore assume that we lost "leadership"?

**Pull request merged**